{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a25940ff-b52d-4dae-bec8-6e6414272cc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:matplotlib.font_manager:Failed to extract font properties from /usr/share/fonts/truetype/noto/NotoColorEmoji.ttf: In FT2Font: Can not load face (unknown file format; error code 0x2)\n",
      "INFO:matplotlib.font_manager:generated new fontManager\n"
     ]
    }
   ],
   "source": [
    "import bioframe as bf\n",
    "import gtfparse\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0a8b8b9-e89d-4bf7-ac7a-81b1ae83edce",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'hgnc_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "# runs for ~ 1 minute\n",
    "gtf = gtfparse.read_gtf('data/gencode.v40.annotation.gtf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0ce70d21-e99a-4d4f-94bd-a9c90dfa125a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = gtf.to_pandas()\n",
    "# seqname is a shit name\n",
    "df = df.rename(columns={'seqname': 'chrom'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "33013bf2-e25e-400f-b8dc-69f684837f73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['', 'ENST00000481143.2', 'ENST00000679437.1', 'ENST00000681026.1',\n",
       "       'ENST00000681989.1', 'ENST00000378723.7', 'ENST00000358062.6',\n",
       "       'ENST00000619831.5', 'ENST00000357033.9', 'ENST00000343523.7',\n",
       "       'ENST00000359836.5', 'ENST00000378707.7', 'ENST00000541735.5',\n",
       "       'ENST00000378677.6', 'ENST00000684130.1', 'ENST00000680701.1',\n",
       "       'ENST00000679850.1', 'ENST00000679641.1', 'ENST00000680355.1',\n",
       "       'ENST00000682322.1', 'ENST00000682600.1', 'ENST00000680162.2',\n",
       "       'ENST00000683675.1', 'ENST00000680961.1', 'ENST00000682238.1',\n",
       "       'ENST00000679482.1', 'ENST00000679669.1', 'ENST00000680557.1',\n",
       "       'ENST00000620040.5', 'ENST00000684072.1', 'ENST00000682746.1',\n",
       "       'ENST00000682207.1', 'ENST00000684350.1', 'ENST00000682769.1',\n",
       "       'ENST00000683709.1', 'ENST00000680768.2', 'ENST00000378702.8',\n",
       "       'ENST00000474231.5', 'ENST00000683509.1', 'ENST00000361471.8',\n",
       "       'ENST00000378680.6', 'ENST00000679706.1', 'ENST00000681153.1',\n",
       "       'ENST00000684103.1', 'ENST00000684342.1', 'ENST00000683995.1',\n",
       "       'ENST00000683503.1', 'ENST00000684247.1', 'ENST00000683957.1',\n",
       "       'ENST00000378705.3', 'ENST00000681654.1', 'ENST00000475732.3',\n",
       "       'ENST00000681334.1', 'ENST00000681646.1', 'ENST00000681870.1',\n",
       "       'ENST00000680216.1', 'ENST00000634285.1', 'ENST00000634315.1',\n",
       "       'ENST00000683450.1', 'ENST00000683851.1', 'ENST00000683117.1',\n",
       "       'ENST00000681839.1', 'ENST00000471779.1', 'ENST00000682135.1',\n",
       "       'ENST00000488902.5', 'ENST00000682183.1', 'ENST00000493412.1',\n",
       "       'ENST00000420596.5', 'ENST00000448370.5', 'ENST00000682899.1',\n",
       "       'ENST00000683985.1', 'ENST00000288447.9', 'ENST00000684165.1',\n",
       "       'ENST00000682071.1', 'ENST00000682924.1', 'ENST00000684292.1',\n",
       "       'ENST00000480751.1', 'ENST00000447523.1', 'ENST00000684237.1',\n",
       "       'ENST00000682870.1', 'ENST00000684660.1', 'ENST00000683309.1',\n",
       "       'ENST00000684056.1', 'ENST00000683658.1', 'ENST00000682437.1',\n",
       "       'ENST00000682584.1', 'ENST00000684357.1', 'ENST00000472681.1',\n",
       "       'ENST00000682307.1', 'ENST00000682439.1', 'ENST00000472266.1',\n",
       "       'ENST00000463609.2'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the transcripts for the DMD gene ?\n",
    "df[df['gene_name'] == 'DMD']['transcript_id'].unique()\n",
    "# Note that there is an empty transcript id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1317847-28be-4e1e-9f47-7184ce82ca8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3182690    DMD\n",
       "Name: gene_name, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What is the name of the gene associated with transcript ENST00000684350.1\n",
    "df[(df['transcript_id'] == 'ENST00000684350.1') & (df['feature'] == 'transcript')]['gene_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "651e7c21-1dd5-4bf2-9ae0-28bd5380734d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many exons does transcript ENST00000684350.1 have\n",
    "df[(df['transcript_id'] == 'ENST00000684350.1') & (df['feature'] == 'exon')].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a514cd18-6e2d-415b-98b3-7614e5eb8dfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "275"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many exons are in the DMD genes ?\n",
    "len(df[(df['gene_name'] == 'DMD') & (df['feature'] == 'exon')]['exon_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "83f2da6e-328b-40f4-8c71-c62b4d912c82",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gamazeps/dev/encode-llama/venv/lib/python3.10/site-packages/bioframe/io/fileops.py:74: ParserWarning: Length of header or names does not match length of data. This leads to a loss of data with index_col=False.\n",
      "  df = pd.read_csv(filepath_or, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>fc</th>\n",
       "      <th>-log10p</th>\n",
       "      <th>-log10q</th>\n",
       "      <th>relSummit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>99392395</td>\n",
       "      <td>99392827</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>306.78095</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr3</td>\n",
       "      <td>194929204</td>\n",
       "      <td>194929707</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>284.09197</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>97559907</td>\n",
       "      <td>97560401</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>284.16880</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>66609310</td>\n",
       "      <td>66609747</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>265.66492</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>66903598</td>\n",
       "      <td>66904095</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>263.27660</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129238</th>\n",
       "      <td>chr17</td>\n",
       "      <td>55794221</td>\n",
       "      <td>55794717</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>5.91739</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.35122</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129239</th>\n",
       "      <td>chr5</td>\n",
       "      <td>39135451</td>\n",
       "      <td>39135947</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>6.62387</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.51343</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129240</th>\n",
       "      <td>chr20</td>\n",
       "      <td>16241885</td>\n",
       "      <td>16242381</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>8.24784</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.96486</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129241</th>\n",
       "      <td>chr2</td>\n",
       "      <td>39231186</td>\n",
       "      <td>39231682</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>6.21278</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.42335</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129242</th>\n",
       "      <td>chr16</td>\n",
       "      <td>84553939</td>\n",
       "      <td>84554435</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>8.06121</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.91185</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129243 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end name  score strand         fc  -log10p  \\\n",
       "0       chr10   99392395   99392827    .   1000      .  306.78095     -1.0   \n",
       "1        chr3  194929204  194929707    .   1000      .  284.09197     -1.0   \n",
       "2        chr8   97559907   97560401    .   1000      .  284.16880     -1.0   \n",
       "3        chr8   66609310   66609747    .   1000      .  265.66492     -1.0   \n",
       "4       chr15   66903598   66904095    .   1000      .  263.27660     -1.0   \n",
       "...       ...        ...        ...  ...    ...    ...        ...      ...   \n",
       "129238  chr17   55794221   55794717    .     78      .    5.91739     -1.0   \n",
       "129239   chr5   39135451   39135947    .     78      .    6.62387     -1.0   \n",
       "129240  chr20   16241885   16242381    .     78      .    8.24784     -1.0   \n",
       "129241   chr2   39231186   39231682    .     78      .    6.21278     -1.0   \n",
       "129242  chr16   84553939   84554435    .     78      .    8.06121     -1.0   \n",
       "\n",
       "        -log10q  relSummit  \n",
       "0       4.96085        230  \n",
       "1       4.96085        234  \n",
       "2       4.96085        242  \n",
       "3       4.96085        238  \n",
       "4       4.96085        252  \n",
       "...         ...        ...  \n",
       "129238  0.35122        248  \n",
       "129239  0.51343        248  \n",
       "129240  0.96486        248  \n",
       "129241  0.42335        248  \n",
       "129242  0.91185        248  \n",
       "\n",
       "[129243 rows x 10 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctcf_peak = bf.read_table(\n",
    "    \"https://www.encodeproject.org/files/ENCFF417KHZ/@@download/ENCFF417KHZ.bed.gz\",\n",
    "     schema=\"narrowPeak\",\n",
    ")   \n",
    "ctcf_peak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fdec808e-20c2-4b80-b97c-93f8adf4b56d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_1</th>\n",
       "      <th>source_1</th>\n",
       "      <th>feature_1</th>\n",
       "      <th>start_1</th>\n",
       "      <th>end_1</th>\n",
       "      <th>score_1</th>\n",
       "      <th>strand_1</th>\n",
       "      <th>frame_1</th>\n",
       "      <th>gene_id_1</th>\n",
       "      <th>gene_type_1</th>\n",
       "      <th>...</th>\n",
       "      <th>chrom_2</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "      <th>name_2</th>\n",
       "      <th>score_2</th>\n",
       "      <th>strand_2</th>\n",
       "      <th>fc_2</th>\n",
       "      <th>-log10p_2</th>\n",
       "      <th>-log10q_2</th>\n",
       "      <th>relSummit_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>31656639</td>\n",
       "      <td>31657135</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>8.80363</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.14717</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>31658342</td>\n",
       "      <td>31658838</td>\n",
       "      <td>.</td>\n",
       "      <td>452</td>\n",
       "      <td>.</td>\n",
       "      <td>20.41173</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>31961756</td>\n",
       "      <td>31962252</td>\n",
       "      <td>.</td>\n",
       "      <td>487</td>\n",
       "      <td>.</td>\n",
       "      <td>19.26018</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.48525</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>32483449</td>\n",
       "      <td>32483945</td>\n",
       "      <td>.</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>14.56001</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.51608</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>32530727</td>\n",
       "      <td>32531223</td>\n",
       "      <td>.</td>\n",
       "      <td>176</td>\n",
       "      <td>.</td>\n",
       "      <td>16.82755</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.18576</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>32687396</td>\n",
       "      <td>32687892</td>\n",
       "      <td>.</td>\n",
       "      <td>202</td>\n",
       "      <td>.</td>\n",
       "      <td>22.98080</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198947.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>33324080</td>\n",
       "      <td>33324576</td>\n",
       "      <td>.</td>\n",
       "      <td>876</td>\n",
       "      <td>.</td>\n",
       "      <td>34.99871</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.96085</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_1 source_1 feature_1   start_1     end_1  score_1 strand_1  frame_1  \\\n",
       "0    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "1    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "2    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "3    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "4    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "5    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "6    chrX   HAVANA      gene  31097677  33339609      NaN        -        0   \n",
       "\n",
       "            gene_id_1     gene_type_1  ... chrom_2   start_2     end_2 name_2  \\\n",
       "0  ENSG00000198947.18  protein_coding  ...    chrX  31656639  31657135      .   \n",
       "1  ENSG00000198947.18  protein_coding  ...    chrX  31658342  31658838      .   \n",
       "2  ENSG00000198947.18  protein_coding  ...    chrX  31961756  31962252      .   \n",
       "3  ENSG00000198947.18  protein_coding  ...    chrX  32483449  32483945      .   \n",
       "4  ENSG00000198947.18  protein_coding  ...    chrX  32530727  32531223      .   \n",
       "5  ENSG00000198947.18  protein_coding  ...    chrX  32687396  32687892      .   \n",
       "6  ENSG00000198947.18  protein_coding  ...    chrX  33324080  33324576      .   \n",
       "\n",
       "  score_2 strand_2      fc_2 -log10p_2 -log10q_2 relSummit_2  \n",
       "0      78        .   8.80363      -1.0   1.14717         248  \n",
       "1     452        .  20.41173      -1.0   4.96085         248  \n",
       "2     487        .  19.26018      -1.0   4.48525         248  \n",
       "3      78        .  14.56001      -1.0   3.51608         248  \n",
       "4     176        .  16.82755      -1.0   4.18576         248  \n",
       "5     202        .  22.98080      -1.0   4.96085         248  \n",
       "6     876        .  34.99871      -1.0   4.96085         248  \n",
       "\n",
       "[7 rows x 35 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Are there any CTCF peaks overlapping with DMD ?\n",
    "bf.overlap(\n",
    "    df[(df['gene_name'] == 'DMD') & (df['feature'] == 'gene')],\n",
    "    ctcf_peak,\n",
    "    how='inner',\n",
    "    suffixes=('_1','_2')\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
